### ISSUE TITLE
repartition dask  dataframe according to frequency doesn't match pandas known frequencies

### ISSUE THREAD

**Issue by MosheVai** [dask/dask#3714]
When repartitioning a dask  dataframe with a date index according to a certain frequency it does not match the partition made with a pandas Grouper object.

example: 
```python
import pandas as pd
lets = [f'{i:02},' for i in range(9)]
dates = pd.date_range('2018-07-01', periods=9, freq='1d')
df = pd.DataFrame({'dates':dates, 'nums': nums}).set_index('dates')
```
when grouping it with a weekly frequency until thursday:
```python
grouper = pd.Grouper(level='dates', freq='W-THU')
df.groupby(grouper).agg({'nums':'sum'})
```
you get the following result

|            | nums                  |
|------------|-----------------------|
| dates      |                       |
| 2018-07-05 | 00,01,02,03,04,       |
| 2018-07-12 | 05,06,07,08, |

however when you repartition a dask dataframe according to the same frequency
and print them you get different partitions. 
```python
import dask.dataframe as dd
ddf = dd.from_pandas(df, npartitions=1).repartition(freq='W-THU')

for i in range(ddf.npartitions):
    print(ddf.get_partition(i).compute())
```
|            | nums |
|------------|------|
| dates      |      |
| 2018-07-01 | 00,  |
| 2018-07-02 | 01,  |
| 2018-07-03 | 02,  |
| 2018-07-04 | 03,  |

|            | nums |
|------------|------|
| dates      |      |
| 2018-07-05 | 04,  |
| 2018-07-06 | 05,  |
| 2018-07-07 | 06,  |
| 2018-07-08 | 07,  |
| 2018-07-09 | 08,  |

and thus resulting in faulty calculations if i do the following:
```python 
grouper = pd.Grouper(level='dates', freq='W-THU')
ddf.map_partitions(lambda df:df.groupby(grouper).agg({'nums':'sum'}), meta={'nums':str}).compute()
``` 
|            | nums         |
|------------|--------------|
| dates      |              |
| 2018-07-05 | 00,01,02,03, |
| 2018-07-05 | 04,          |
| 2018-07-12 | 05,06,07,08, |

**Comment by MosheVai** [comment]
i dont know if this should be in a separate issue but there are times when weekly partitions contain 8 days.
example: 
```python
nums = [f'{i:02},' for i in range(0,13)]
dates = pd.date_range('2018-07-01', periods=13, freq='1d')

ddf = dd.from_pandas(df, npartitions=1).repartition(freq='W-FRI')

for i in range(ddf.npartitions):
    print(ddf.get_partition(i).compute())
``` 
Note that this only happens in the last partition (i have tried this with more data).

|            | nums |
|------------|------|
| dates      |      |
| 2018-07-01 | 00,  |
| 2018-07-02 | 01,  |
| 2018-07-03 | 02,  |
| 2018-07-04 | 03,  |
| 2018-07-05 | 04,  |


|            | nums |
|------------|------|
| dates      |      |
| 2018-07-06 | 05,  |
| 2018-07-07 | 06,  |
| 2018-07-08 | 07,  |
| 2018-07-09 | 08,  |
| 2018-07-10 | 09,  |
| 2018-07-11 | 10,  |
| 2018-07-12 | 11,  |
| 2018-07-13 | 12,  |

**Comment by TomAugspurger** [comment]
Thanks for the report. Something is going awry in https://github.com/dask/dask/blob/d8a4f3be9a6ec685261ed043a45ecc019055ec0d/dask/dataframe/core.py#L4432.

I think this is perhaps hard to support. Dask's metadata requires that the divisions be the minimum value in each partition and the max value of the last partition. When a user passes an anchored offset like `W-THU`, we're stuck because the first / last value may not be on that anchor.

To summarize, I think this example should raise on the `.repartition(freq='W-THU')` when we can't actually create divisions with that frequency (i.e. the current dataset doesn't end and begin on a Thursday. Does that make sense?
